pytorch_lightning/strategies/collaborative.py

```diff
@@ -0,0 +1,513 @@
+        matchmaking_time: float = 5.0,
+        averaging_timeout: float = 30.0,
+        verbose: bool = False,
+        averager_opts: Optional[Dict] = None,
+        host_maddrs: Optional[List] = None,
+        initial_peers: Optional[Union[str, List]] = None,
+        endpoint: Optional[bool] = None,
+        peer_endpoint: Optional[str] = None,
+        persistent: bool = True,
+        host: Optional[str] = None,
+        port: Optional[int] = None,
+        retry_endpoint_attempts: int = 5,
+        retry_endpoint_sleep_duration: int = 5,
+        **optimizer_kwargs: Any,
+    ):
+        """Provides capabilities to train using the Hivemind Library, training collaboratively across the internet
+        with unreliable machines. For more information, `refer to the docs <https://pytorch-
+        lightning.readthedocs.io/en/latest/strategies/collaborative_training.html>`__.
+
+        .. warning:: ``CollaborativeStrategy`` is experimental and subject to change.
+
+        Arguments:
+
+            target_batch_size: When training, the batch size to accumulate to before running a step. The larger this
+                batch size, the more work can be done asynchronously without communication.
+
+            run_id: A unique identifier of this training run, used as a common prefix for all DHT keys.
+
+            batch_size: The local batch size per process. If not provided,
+                we infer this from the first batch of data passed in at training.
+
+            delay_state_averaging: If enabled (default), average parameters and extra tensors in a background thread;
+                if set to False, average parameters synchronously within the corresponding hivemind.Optimizer.step call.
+
+            delay_optimizer_step: Run optimizer in background, apply results in future .step.
+                requires `offload_optimizer`.
+
+            delay_grad_averaging: Average gradients in background; requires offload_optimizer and delay_optimizer_step.
+
+            offload_optimizer: Offload the optimizer to host memory, saving GPU memory for parameters and gradients.
+
+            reuse_grad_buffers: Use model's .grad buffers for gradient accumulation.
+                This is more memory efficient, but it requires that we do not call `zero_grad` in the `LightningModule`.
+
+            scheduler_fn: callable(optimizer) -> PyTorch LRScheduler or a pre-initialized PyTorch scheduler.
+                When using `offload_optimizer`/`delay_optimizer_step`/`delay_state_averaging` a scheduler_fn is required
+                instead of passing a scheduler to the `pl.LightningModule`. This is because the optimizer is re-created
+                and the scheduler needs to be re-created as well.
+
+            matchmaking_time: When looking for group, wait for peers to join for up to this many seconds.
+                Increase if you see "averaged gradients with N peers" where N is below 0.9x on >=25% of epochs.
```

Comment:
Is this a heuristic? Is that why you dont' do it automatically?

Reply:
It's tricky as you do not want to hang for too long (if processes are on a very unstable connection). automating some of these variables is a next step from this PR, this just exposes vars that are important for hivemind.